pkg/envoy/ads/stream.go

```diff
@@ -136,82 +138,104 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), discoveryRequest.ResourceNames)
 		return true
 	}
 
-	// The version of the config received along with the DiscoveryRequest (ackVersion)
-	// is what the Envoy proxy may be acknowledging. It is acknowledging
-	// and not requesting when the ackVersion is <= what we last sent.
-	// It is possible however for a proxy to have a version that is higher
-	// than what we last sent. (Perhaps the control plane restarted.)
-	// In that case we want to make sure that we send new responses with
-	// VersionInfo incremented starting with the version which the proxy last had.
-	lastVersion = proxy.GetLastSentVersion(typeURL)
-	if requestVersion > lastVersion {
-		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Higher version on request %s, req ver: %d - last ver: %d. Updating to match latest.",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestVersion, lastVersion)
+	// Handle reconnection case for a proxy that already has some configuration applied
+	lastNonce = proxy.GetLastSentNonce(typeURL)
+	if lastNonce == "" {
+		// This is a new connection case, where the incoming nonce is not empty. This is the case of a proxy that lost connection to its control plane
+		// and connected back to a control plane.
+		// Version applied is going to be X, we will set our version to be also X, and trigger a response. This will make
+		// this control plane to generate version X+1 for this proxy, thus keeping linearity between versions even if the proxy
+		// moves around different control planes, and updating the resources to the SotW of this control plane.
+		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Request type %s nonce %s for a proxy we didn't yet issue a nonce for. Updating version to %d.",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestNonce, requestVersion)
 		proxy.SetLastSentVersion(typeURL, requestVersion)
+		proxy.SetLastAppliedVersion(typeURL, requestVersion)
 		return true
 	}
 
-	// Compare Nonces
-	// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce, if the
-	// nonce is non-empty.
-	lastNonce = proxy.GetLastSentNonce(typeURL)
+	// Handle regular proto (nonce based) from now on
+	// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce.
 	if requestNonce != lastNonce {
 		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Ignoring request for %s non-latest nonce (request: %s, current: %s)",
 			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestNonce, lastNonce)
 		return false
 	}
 
+	// Nonces match
+	// At this point, there is no error and nonces match, it is guaranteed an ACK with last sent version.
+	proxy.SetLastAppliedVersion(typeURL, requestVersion)
```

Comment:
Interesting nuance here - we set the LastAppliedVersion when nonces match and not when it is empty ✔
Which makes sense and I wonder why we had to do it the way it was done before this PR  🤔